### PR TITLE
Call gas work

### PR DESCRIPTION
**Pull request by OlivierBBB**

### CODE REVIEW

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/CallSection.java [hunk 1]
```diff
@@ -446,6 +446,9 @@ public void resolveAtContextReEntry(Hub hub, CallFrame frame) {
         checkState(successBit);
         scenarioFragment.setScenario(CALL_EOA_SUCCESS_WONT_REVERT);
         emptyCodeFirstCoupleOfAccountFragments(hub);
+        long gasAfterCall = frame.frame().getRemainingGas();
+        commonValues.gasNext(gasAfterCall);
+        hub.currentFrame().lastValidGasNext(gasAfterCall);
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/CallSection.java:453]: For EOA's and SMC's we grab the `gasRemaining` in the frame when it resumes execution. This frees us from having to compute it by hand and fixes many of the gas issues we were having with CALL-type instructions.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/CallSection.java [hunk 2]
```diff
@@ -455,6 +458,9 @@ public void resolveAtContextReEntry(Hub hub, CallFrame frame) {
           scenarioFragment.setScenario(CALL_PRC_FAILURE);
         }
         emptyCodeFirstCoupleOfAccountFragments(hub);
+        long gasAfterCall = frame.frame().getRemainingGas();
+        commonValues.gasNext(gasAfterCall);
+        hub.currentFrame().lastValidGasNext(gasAfterCall);
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/CallSection.java:465]: same

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/StpCall.java [hunk 3]
```diff
@@ -135,15 +138,12 @@ private long computeGasPaidOutOfPocketForCreates() {
...
-    boolean toIsWarm = warm;
     long upfrontGasCost = memoryExpansionGas;
-    final boolean callWouldLeadToAccountCreation = isCALL && nonzeroValueTransfer && !exists;
-    if (nonzeroValueTransfer) upfrontGasCost += GAS_CONST_G_CALL_VALUE;
-    if (toIsWarm) upfrontGasCost += GAS_CONST_G_WARM_ACCESS;
-    else upfrontGasCost += GAS_CONST_G_COLD_ACCOUNT_ACCESS;
-    if (callWouldLeadToAccountCreation) upfrontGasCost += GAS_CONST_G_NEW_ACCOUNT;
+    upfrontGasCost += nonzeroValueTransfer() ? GAS_CONST_G_CALL_VALUE : 0;
+    upfrontGasCost += warm ? GAS_CONST_G_WARM_ACCESS : GAS_CONST_G_COLD_ACCOUNT_ACCESS;
+    upfrontGasCost += callWouldLeadToAccountCreation() ? GAS_CONST_G_NEW_ACCOUNT : 0;
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/StpCall.java:144]: **bugfix** + streamlining of the computation.

The bug was that the transafer value of 9000 was only paid for `CALL`, while both `CALL` and `CALLCODE` can trigger it.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/runtime/callstack/CallFrame.java [hunk 4]
```diff
@@ -94,7 +94,7 @@ public int getCodeFragmentIndex(Hub hub) {
   @Getter @Setter private MemoryRange outputDataRange = MemoryRange.EMPTY; // set at exit time
 
   @Getter private boolean executionPaused = false;
-  @Getter private long lastValidGasNext = 0;
+  @Getter @Setter private long lastValidGasNext = 0;
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/runtime/callstack/CallFrame.java:96]: Used to set the GAS_NEXT after the fact for EOA and PRC calls.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/CallSection.java (hunk 2 above)
Comment (letypequividelespoubelles) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/CallSection.java:465]: doesn't at this point hub.currentFrame == frame ?

**Review by letypequividelespoubelles — APPROVED** [review]
LGTM, made a little formatting commit + small question

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/CallSection.java (hunk 2 above)
Comment (letypequividelespoubelles) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/CallSection.java:465]: I mean the frame of resolveAtContextReEntry(Hub hub, CallFrame frame)

Comment (OlivierBBB) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/CallSection.java:465]: I'm not sure ...

**Pull request merged**